File: 18. Backend Scaling and Performance Engineering-part-1/code/typescript/example_5.ts
import { Pool } from 'pg';

const db = new Pool({
    connectionString: process.env.DATABASE_URL,
    max: 25,               // max open connections
    idleTimeoutMillis: 60000, // close idle connections after 1 min
    // Node pg doesn't have an exact maxLifetime equivalent natively in connection pool,
    // but max and idleTimeoutMillis are the main knobs.
});

db.on('error', (err) => {
    console.error('Unexpected error on idle client', err);
    process.exit(-1);
});

File: 18. Backend Scaling and Performance Engineering-part-1/code/rust/example_5.rs
use sqlx::postgres::PgPoolOptions;
use std::time::Duration;

#[tokio::main]
async fn main() -> Result<(), sqlx::Error> {
    let db = PgPoolOptions::new()
        .max_connections(25)
        .min_connections(10) // max idle connections kept ready
        .max_lifetime(Duration::from_secs(5 * 60)) // recycle connections after 5 min
        .idle_timeout(Duration::from_secs(60)) // close idle connections after 1 min
        .connect("postgres://...")
        .await?;
        
    Ok(())
}
